### PR TITLE
Eq 320 piping

### PR DESCRIPTION
**Pull request by weapdiv-david**
### Changes

Pipes exercise start and end dates into question text
### How to test

1) Check out the branch
2) Run the unit tests
3) Run the app and check the date period question
4) Create a new token using dev mode with new dates and check that the same questionnaire shows the dates in your token
### Who can test

Anybody except @weapdiv-david

### CODE REVIEW

File: app/data/1_0203.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
                             {
                               "id": "6cc86b54-330c-4465-99b2-34cc7073dc2c",
                               "title": "What are the dates of the sales period you are reporting for?",
-                              "description": "If possible, this should be for the period 3rd April 2016 to 30th April 2016.",
+                              "description": "If possible, this should be for the period {exercise.start_date:%-d %B %Y} to {exercise.end_date:%-d %B %Y}.",
```
Comment (LJBabbage) [inline app/data/1_0203.json:28]: We need to escape { if anyone tries to use it in the text it blows up

File: app/piping/plumber.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import re
+
+
+class Plumber(object):
+    def __init__(self, context):
+        self._context = context
+
+    def plumb_item(self, item):
+        interesting_properties = self.get_properties_of_interest(item)
```
Comment (LJBabbage) [inline app/piping/plumber.py]: Not sure what interesting_properties means, can you change the name to something more descriptive

File: app/data/1_0203.json (hunk 1 above)
Comment (collisdigital) [inline app/data/1_0203.json:28]: Consider adding a test that checks this situation.

File: app/renderer/renderer.py [hunk 3]
```diff
@@ -112,47 +136,53 @@ def _render_navigation_meta(self):
...
+                    response.warnings = None
 
     def _augment_questionnaire(self):
         errors = OrderedDict()
         warnings = OrderedDict()
 
         # loops through the Schema and get errors and warnings in order
+        # augments each item in the schema as required
         for group in self._schema.groups:
+            self._plumber.plumb_item(group)
```
Comment (collisdigital) [inline app/renderer/renderer.py]: As a minimum these steps (lines 161-166) can be factored out onto a function and called for each of the repeats in the nested loops below. Further refactoring may be possible beyond this to clean this up.

File: app/metadata/metadata_store.py [hunk 4]
```diff
@@ -76,16 +81,20 @@ def get_form_type(self):
         return self.get(MetaDataConstants.FORM_TYPE)
 
     def get_ref_p_start_date(self):
-        return self.get(MetaDataConstants.REF_P_START_DATE)
+        if self.user is None:
+            return None
+        return datetime.strptime(self.get(MetaDataConstants.REF_P_START_DATE), "%Y-%m-%d")
```
Comment (warrenbailey) [inline app/metadata/metadata_store.py]: I'd like to understand in what situations self.user is None as it shouldn't be.

Also you're already doing a self.user check in  self.get. I wonder if this would be better written:

```
def get_ref_p_start_date(self):
        start_date = self.get(MetaDataConstants.REF_P_START_DATE)       
        if start_date:
            return datetime.strptime(start_date, "%Y-%m-%d")
```

File: app/metadata/metadata_store.py [hunk 5]
```diff
@@ -40,14 +41,18 @@ def __init__(self, user=None):
             self.user = current_user
 
     def store(self, key, value):
-        data = self.user.get_questionnaire_data()
-        data[key] = value
+        if self.user is not None:
```
Comment (warrenbailey) [inline app/metadata/metadata_store.py]: I wonder if we re-write this class and initialize all the variables on construction. That way it could be immutable and we wouldn't need these checks.

File: app/renderer/renderer.py [hunk 6]
```diff
@@ -13,6 +13,31 @@ def __init__(self, schema, response_store, validation_store, navigator, metadata
...
+
+        try:
+            start_date = self._metadata.get_ref_p_start_date()
+            end_date = self._metadata.get_ref_p_end_date()
+        except:
+            pass
+
+        context = {
+            "exercise": ContextEntry({
+                # "start_date": '{dt.day} {dt:%B} {dt.year}'.format(dt=datetime.strptime(self._metadata.get_ref_p_start_date(), "%Y-%m-%d")),
```
Comment (warrenbailey) [inline app/renderer/renderer.py]: Lets just remove the commented out code if its not needed.

File: app/renderer/renderer.py [hunk 7]
```diff
@@ -13,6 +13,31 @@ def __init__(self, schema, response_store, validation_store, navigator, metadata
         self._navigator = navigator
         self._metadata = metadata
 
+        # Create an object from a dict
+        class ContextEntry:
+            def __init__(self, entries):
+                self.__dict__ = entries
```
Comment (warrenbailey) [inline app/renderer/renderer.py]: You have exactly the same class called ObjectFromDict in test_plumber.py. Maybe refactor both into the app.utilities package.

File: app/renderer/renderer.py [hunk 8]
```diff
@@ -13,6 +13,31 @@ def __init__(self, schema, response_store, validation_store, navigator, metadata
...
+
+        start_date = None
+        end_date = None
+
+        try:
+            start_date = self._metadata.get_ref_p_start_date()
+            end_date = self._metadata.get_ref_p_end_date()
+        except:
+            pass
```
Comment (warrenbailey) [inline app/renderer/renderer.py:24]: I don't really understand this try/except block. Either catch the exception and log it or at the very least comment whats going on as to me this looks like it will just quietly swallow all errors.

File: app/piping/plumber.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        for interesting_property in interesting_properties:
+            if self._needs_plumbing(item, interesting_property):
+                self._plumb(item, interesting_property)
+
+    def get_properties_of_interest(self, item):
+        if hasattr(item, 'templatable_properties'):
```
Comment (warrenbailey) [inline app/piping/plumber.py]: Would it be possible to have templatable_propertes value on all aspects of the schema? We could set it to [] to start with?

File: app/piping/plumber.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            return item.templatable_properties
+        else:
+            return []
+
+    def _needs_plumbing(self, item, interesting_property):
+        value = getattr(item, interesting_property)
```
Comment (warrenbailey) [inline app/piping/plumber.py]: maybe call this variable 'template' to match the one below

File: app/piping/plumber.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import re
+
+
+class Plumber(object):
+    def __init__(self, context):
```
Comment (warrenbailey) [inline app/piping/plumber.py:5]: I'm finding this class a little hard to follow, maybe add some comments to explain whats going on.

File: app/piping/plumber.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import re
+
+
```
Comment (warrenbailey) [inline app/piping/plumber.py:3]: I think this class needs debug logging.

File: app/renderer/renderer.py (hunk 8 above)
Comment (warrenbailey) [inline app/renderer/renderer.py:24]: Could still do with explaining this.

File: app/piping/plumber.py (hunk 11 above)
Comment (warrenbailey) [inline app/piping/plumber.py:5]: With the variable name change it's definitely better, but a bit of documentation wouldn't hurt

**Comment by warrenbailey** [comment]
Looks good now, a couple of minor things but my only comment now is it has no logging. I really think some debug logging in here, especially the plumber class, would be really useful. Even if it just tells us what properties are being replace with what values.

**Comment by warrenbailey** [comment]
LGTM 👍
